File: 516_longest_palindromic_subsequence.java
class Solution {
    // use memorization to record the number of characters in subsequence
    // given left and right indexes
    // n = len(s) => Time Complexity: O(n^2), Space Complexity: O(n^2);
    public int longestPalindromeSubseq(String s) {
        Integer[][] mem = new Integer[s.length()][s.length()];
        return findLongestPalindromeSeq(0, s.length() - 1, s, mem);
    }

    public int findLongestPalindromeSeq(int left, int right, String s, Integer[][] mem) {
        if (left == right) return 1;
        if (left > right) return 0;
        if (mem[left][right] != null) return mem[left][right];

        return mem[left][right] = ((s.charAt(left) == s.charAt(right)) ?
            2 + findLongestPalindromeSeq(left + 1, right - 1, s, mem) :
            Math.max(findLongestPalindromeSeq(left + 1, right, s, mem), findLongestPalindromeSeq(left, right - 1, s, mem)));
    }
}

File: python/516_longest_palindromic_subsequence.py
class Solution:
    # Time Complexity: O(n ^ 2)
    # Space Complexity: O(n ^ 2)
    def longestPalindromeSubseq(self, s: str) -> int:
        # palindromic -> symmetric
        n = len(s)
        dp = [[0 for _ in range(n)] for _ in range(n)]

        for i in range(n - 1, -1, -1):
            dp[i][i] = 1
            for j in range(i + 1, n):
                if s[i] == s[j]:
                    # plus 2: 向外擴展兩個字元，故加 2
                    dp[i][j] = dp[i + 1][j - 1] + 2
                else:
                    dp[i][j] = max(dp[i + 1][j], dp[i][j - 1])
        return dp[0][n - 1]
